### ISSUE TITLE
"Link Another Identity" doesn't; it may create a second account

### ISSUE THREAD

**Issue by stephen-smith** [davido/gerrit-oauth-provider#62]
Brand new installation, but OpenID is impossible to use the way it's present in stock gerrit, so I went with OAuth-only.

Created first account using GitHub, where I use email bss@iguanasuicide.net.

Went to settings and tried to "Link Another Account".

OAuth'ed with Google, where I use email Boyd.Stephen.Smith.Jr@GMail.com

I'm now logged in at a separate account, and neither is linked with the other!

Quite surprising!  If the email addresses have to be the same... I'd expect "Link"ing to fail with an informative error message, not create a new account!

**Comment by stephen-smith** [comment]
Then, I can log off entirely, Log in using Google OAuth, then try and Link with my BitBucket OAuth and the link "succeeds", but it's linked to by GitHub OAuth (!) and *not* to my Google OAuth.  BTW, I magically changed logged in user again.

Now, attempting to link either GitHub or BitBucket when I'm logged in as my Google account, results in:
```
[2016-09-15 18:30:21,474] [HTTP-61] ERROR com.google.gerrit.httpd.auth.oauth.OAuthSession : Unable to authenticate user "com.google.gerrit.extensions.auth.oauth.OAuthUserInfo@30f3c140"
com.google.gerrit.server.account.AccountException: Identity in use by another account
        at com.google.gerrit.server.account.AccountManager.link(AccountManager.java:354)
        at com.google.gerrit.httpd.auth.oauth.OAuthSession.authenticateWithLinkedIdentity(OAuthSession.java:201)
        at com.google.gerrit.httpd.auth.oauth.OAuthSession.authenticateAndRedirect(OAuthSession.java:138)
        at com.google.gerrit.httpd.auth.oauth.OAuthSession.login(OAuthSession.java:104)
        at com.google.gerrit.httpd.auth.oauth.OAuthWebFilter.doFilter(OAuthWebFilter.java:109)
        at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:82)
        at com.google.gwtexpui.server.CacheControlFilter.doFilter(CacheControlFilter.java:73)
        at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:82)
        at com.google.gerrit.httpd.RunAsFilter.doFilter(RunAsFilter.java:117)
        at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:82)
        at com.google.gerrit.httpd.RequireSslFilter.doFilter(RequireSslFilter.java:68)
        at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:82)
        at com.google.gerrit.httpd.AllRequestFilter$FilterProxy$1.doFilter(AllRequestFilter.java:136)
        at com.google.gerrit.httpd.AllRequestFilter$FilterProxy.doFilter(AllRequestFilter.java:105)
        at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:82)
        at com.google.gerrit.httpd.RequestContextFilter.doFilter(RequestContextFilter.java:75)
        at com.google.inject.servlet.FilterChainInvocation.doFilter(FilterChainInvocation.java:82)
        at com.google.inject.servlet.ManagedFilterPipeline.dispatch(ManagedFilterPipeline.java:119)
        at com.google.inject.servlet.GuiceFilter$1.call(GuiceFilter.java:133)
        at com.google.inject.servlet.GuiceFilter$1.call(GuiceFilter.java:130)
        at com.google.inject.servlet.GuiceFilter$Context.call(GuiceFilter.java:203)
        at com.google.inject.servlet.GuiceFilter.doFilter(GuiceFilter.java:130)
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:585)
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:221)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
        at org.eclipse.jetty.server.handler.RequestLogHandler.handle(RequestLogHandler.java:95)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
        at org.eclipse.jetty.server.Server.handle(Server.java:499)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:310)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
        at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:540)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
        at java.lang.Thread.run(Thread.java:745)
```

Emphasis on the "Identity in use by another account" part.

I can understand not being able to auto-merge the accounts but a page with just "Forbidden" would be quite mysterious if I couldn't access the error logs.  Maybe a nicer error message could be put into place?

**Comment by davido** [comment]
Yeah, it would be a nice enhancement.

**Comment by ceejatec** [comment]
We are experiencing this as well, with Gerrit 2.13.5. It appears to be quite consistent. If you log in to an existing account, and then attempt to Link Identity and use GitHub auth, it will update the existing account ID in ACCOUNT_EXTERNAL_IDS with the new GitHub ID, but then attempt to create a new account anyway with that same GitHub ID, resulting in the stack trace above and a "Forbidden" for the user.

A coworker said that as far as he could tell, the problem was that the secondary index on ACCOUNT_EXTERNAL_IDS was not getting updated. I don't know enough about the underlying code to say whether that makes sense. I also don't know for sure if this is a Gerrit bug or a gerrit-oauth-provider bug, but it's quite serious - when somebody gets into this situation there appears to be no way to rectify it, even by mucking around with gsql. We may have to revert to 2.12 to avoid the problem.

**Comment by davido** [comment]
It was long time, I looked into it. Can you try to describe the problem and provide exact the steps how to reproduce it?

**Comment by davido** [comment]
Just noticed, that this changes wasn't yet merged in stabe-2.13 branch: [1].

* [1] https://gerrit-review.googlesource.com/99290

**Comment by davido** [comment]
I think this was fixed meantime. Feel free to re-open.
